portfolio/src/main/java/com/google/sps/data/TimeRangeGroup.java

```diff
@@ -0,0 +1,97 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/** Models a collection of Time Ranges with methods for add and delete. */
+public class TimeRangeGroup {
```

Comment:
I have a few suggestions.

First, I would suggest that you use an interface, which can have multiple implementations. The interface can be named TimeRangeGroup, and maybe the initial implementation can be named like ArrayListTimeRangeGroup, given that it's backed by an array-list internally. But in the future there can be other implementations of the interface, like LinkedListTimeRangeGroup or BinaryTreeTimeRangeGroup.

Here are the methods that I see currently, for reference:
public void addTimeRange(TimeRange newTimeRange);
public boolean equals(Object other);
public List<TimeRange> getAllTimeRanges();
public List<TimeRange> deleteEntireTimeRange(TimeRange timeRange);
public List<TimeRange> deletePartOfTimeRange(
    TimeRange timeRangeOriginal, TimeRange timeRangeToDelete);

1. Given that you are trying to use this data structure to track the free time ranges, I think you may want to implement some merging behavior when it comes to addTimeRange. For example, suppose that there was one time range internally, like 1pm-2pm today. And if the user adds another time range from 1:30pm-4pm, then it's equivalent to one time range from 1pm-4pm (free time).

2. I think it can be worthwhile to add another method like hasTimeRange. If for example internally the class was keeping track of 1pm-2pm and 3pm-4pm, and hasTimeRange is called with 3pm-3:30pm, then it would return true. If it was called with 1pm-4pm, then it would return false, because there is a gap (2pm-3pm).

3. I think deleteEntireTimeRange and deletePartOfTimeRange can be merged into deleteTimeRange, which takes some time range. And the method can throw an exception if hasTimeRange is false. For example, suppose internally the class was keeping track of 1pm-2pm and 3pm-4pm. If delete time range is called with 1pm-2pm, then just 3pm-4pm remains. If the same method is called with 1:30pm-1:45pm, then we are left with 1pm-1:30pm, 1:45pm-2pm, and 3pm-4pm. This delete method can be used when a task gets scheduled.

4. I think getAllTimeRanges should return the merged, disjoint time ranges, probably sorted by time range start times. I don't think it should just return the time ranges that were added. For example, suppose two time ranges were added: 1pm-2pm, and then 10am-5pm. When getAllTimeRanges is called, it should return 10am-5pm, rather than [1pm-2pm, 10am-5pm].